### ISSUE TITLE
graph.items() should raise an exception if the node isn't a list

### ISSUE THREAD

**Issue by vporton** [RDFLib/rdflib#841]
`graph.items()` should raise an exception if the node isn't a list.

As a temporary measure you can add

        if not graph.value(self, RDF.first):
            raise ...

**Comment by vporton** [comment]
Ugh, I was wrong: This would also raise an exception on an empty list.

Not sure what to do.

**Comment by ashleysommer** [comment]
I've been encountering this same issue, and I too don't know what the ideal solution is. I'm not sure if there is any way of checking the node before traversing it, to know if it is a valid list or not.

The way I am working at the moment, I am just treating a non-a-list and a empty-list as the same thing in my application logic.

**Comment by CaptSolo** [comment]
@ashleysommer how does one represent an empty-list as an RDF collection?

**Comment by vporton** [comment]
@CaptSolo 
```
L rdf:rest rdf:nil
```

In Turtle empty list can also be made as `()`.

**Comment by CaptSolo** [comment]
@vporton thanks for the hint.

Turns out an empty list is represented by `rdf:nil`: https://www.w3.org/TR/rdf-schema/#ch_nil

`.list()` could be modified to distinguish between a non-list (= throw an exception) and an empty list. Would it be safe to do so = will that not break existing implementations that rely on this behavior?

**Comment by jiggylepcha** [comment]
Hey I want to solve this PR . As `list` is a keyword in python why is it acting like a variable in graph.items()?

**Comment by vporton** [comment]
> Hey I want to solve this PR . As `list` is a keyword in python why is it acting like a variable in graph.items()?

First, it is an issue not a PR. Which PR do you refer to? I don't understand (or don't remember) where `list` is used like a variable.

Second, `list` is not a keyword. It can be used as a variable if one wishes (however, this is not recommended, because the user may expect that it is the list type, so you could rename the variable).

**Comment by jiggylepcha** [comment]
Hey I raised the exception when node in `graph.items()` is not a list but the Travis CI build is failing because it is raising this exception.

**Comment by nicholascar** [comment]
Hi all, we - maintainers - have reviewed this Issue and PR #1078. We've decided not to make any change here as we think that existing behaviour - the `.items()` function stops `yield`ing if no `first` value is found, is correct. So, currently, a Collection will `yeild` `item()` values and empty Collections and non-Collections won't `yield` anything. 

If we allowed this change, we would have expensive checking for all sorts of things going on.
